controllers/perconaserverformysql_controller.go

```diff
@@ -0,0 +1,305 @@
+// PerconaServerForMYSQLReconciler reconciles a PerconaServerForMYSQL object
+type PerconaServerForMySQLReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+//+kubebuilder:rbac:groups=ps.percona.com,resources=perconaserverformysqls;perconaserverformysqls/status;perconaserverformysqls/finalizers,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=pods;configmaps;services;secrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=apps,resources=statefulsets,verbs=get;list;watch;create;update;patch;delete
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *PerconaServerForMySQLReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&apiv2.PerconaServerForMySQL{}).
+		Complete(r)
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the PerconaServerForMYSQL object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *PerconaServerForMySQLReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	nn := req.NamespacedName
+	l := log.FromContext(ctx).
+		WithName("PerconaServerForMySQL").
+		WithValues("name", nn.Name, "namespace", nn.Namespace)
+
+	cr, err := r.loadCR(ctx, nn)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, errors.Wrap(err, "load CR")
+	}
+
+	if err := r.doReconcile(ctx, l, cr); err != nil {
+		return ctrl.Result{RequeueAfter: 5 * time.Second}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (r *PerconaServerForMySQLReconciler) doReconcile(
+	ctx context.Context,
+	log logr.Logger,
+	cr *apiv2.PerconaServerForMySQL,
+) error {
+	if err := r.reconcileUserSecrets(ctx, cr); err != nil {
+		return errors.Wrap(err, "users secret")
+	}
+	if err := r.createTLSSecret(ctx, cr); err != nil {
+		return errors.Wrap(err, "TLS secret")
+	}
+	if err := r.reconcileDatabase(ctx, cr); err != nil {
+		return errors.Wrap(err, "database")
+	}
+	if err := r.reconcileOrchestrator(ctx, cr); err != nil {
+		return errors.Wrap(err, "orchestrator")
+	}
+	if err := r.reconcileReplication(ctx, cr); err != nil {
+		return errors.Wrap(err, "replication")
+	}
+
+	return nil
+}
+
+func (r *PerconaServerForMySQLReconciler) loadCR(ctx context.Context, nn types.NamespacedName) (*apiv2.PerconaServerForMySQL, error) {
+	o, err := k8s.GetObjectWithDefaults(ctx, r.Client, nn, &apiv2.PerconaServerForMySQL{})
+	if err != nil {
+		return nil, err
+	}
+
+	return o.(*apiv2.PerconaServerForMySQL), nil
+}
```

Comment:
function has no sense, I prefer to run `GetObjectWithDefaults` directly or put `GetObjectWithDefaults` logic inside this function